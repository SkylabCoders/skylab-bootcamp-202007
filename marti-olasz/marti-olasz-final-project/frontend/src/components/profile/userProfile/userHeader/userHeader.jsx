import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';

import './userHeader.scss';
import newToast from '../../../../toasts/newToasts';

import isAnEmptyObject from '../../../../isAnEmptyObject';

import {
  userEdit,
  sendUserEditInfo
} from '../../../../redux/actions/authActions';
import { uploadImage, clearImage } from '../../../../redux/actions/infoActions';

import Gear from '@material-ui/icons/Settings';
import Save from '@material-ui/icons/Save';

function ProfileHeader({ user, editInfo, image, dispatch }) {
  const [photo, setPhoto] = useState();
  const [banner, setBanner] = useState();

  const [photoInput, setPhotoInput] = useState(null);
  const [bannerInput, setBannerInput] = useState(null);

  useEffect(() => {
    if (image.identifier === 'profile-photo') {
      setPhoto(image.path);
    } else if (image.identifier === 'banner') {
      setBanner(image.path);
    }
  }, [image]);

  useEffect(() => {
    if (
      (photo === null || typeof photo === 'string') &&
      (banner === null || typeof banner === 'string')
    ) {
      const data = { ...editInfo };
      if (photo) data.photo = photo;
      if (banner) data.banner = banner;
      dispatch(sendUserEditInfo(user._id, data));
      dispatch(userEdit({}));
      dispatch(clearImage());
      newToast('Information saved correctly');
    }
  }, [photo, banner]);

  return (
    <section className='user-header'>
      <img
        src={user.photo}
        alt='user'
        className='user-header__photo'
        onClick={() => {
          if (!isAnEmptyObject(editInfo)) photoInput.click();
        }}
      />
      <input
        type='file'
        name='file'
        style={{ display: 'none' }}
        ref={(fileInput) => setPhotoInput(fileInput)}
        onChange={(event) => setPhoto(event.target.files[0])}
      />
      <div className='user-header__banner'>
        <img
          src={user.banner}
          alt='Banner'
          className='banner__img'
          onClick={() => {
            if (!isAnEmptyObject(editInfo)) bannerInput.click();
          }}
        />

        <input
          type='file'
          name='file'
          style={{ display: 'none' }}
          ref={(fileInput) => setBannerInput(fileInput)}
          onChange={(event) => setBanner(event.target.files[0])}
        />
      </div>
      <div className='user-header__edit'>
        <Gear
          className={`edit__gear ${
            editInfo.user === undefined ? 'white' : 'orange'
          }`}
          onClick={(event) =>
            dispatch(
              userEdit(
                editInfo.user === undefined
                  ? { user: user.user, bio: user.bio }
                  : {}
              )
            )
          }
        />
        {!isAnEmptyObject(editInfo) && (
          <Save
            className='edit__save'
            onClick={(event) => {
              if (photo) {
                const img = new FormData();
                img.append('file', photo);
                dispatch(
                  uploadImage(
                    `user/${user._id}/profile-photo`,
                    img,
                    'profile-photo'
                  )
                );
              } else setPhoto(null);

              if (banner) {
                const img = new FormData();
                img.append('file', banner);
                dispatch(uploadImage(`user/${user._id}/banner`, img, 'banner'));
              } else setBanner(null);
            }}
          />
        )}
      </div>
    </section>
  );
}

function mapStateToProps(state) {
  return {
    editInfo: state.authReducer.editInfo,
    user: state.authReducer.user,
    image: state.infoReducer.image
  };
}

export default connect(mapStateToProps)(ProfileHeader);
export { ProfileHeader };
