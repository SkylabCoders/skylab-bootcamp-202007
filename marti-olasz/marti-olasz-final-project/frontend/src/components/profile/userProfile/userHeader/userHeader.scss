$color-detail: #f55110;
$max-banner-width: 600px;

.user-header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  height: 200px;

  padding-bottom: 13px;
  border-bottom: 4px solid $color-detail;
  margin-bottom: 0px;

  .user-header__photo {
    width: 200px;
  }
  .user-header__banner {
    display: none;

    flex: 1;
    justify-content: center;

    margin-left: 30px;

    .banner__img {
      max-width: $max-banner-width;
    }
  }
  .user-header__edit {
    display: flex;
    align-items: flex-end;
    justify-content: center;

    width: 100px;
    margin-left: 30px;

    .edit__gear {
      font-size: 40px;
      color: $color-detail;
    }

    .edit__save {
      font-size: 40px;
      color: white;
    }

    .white {
      color: white;
    }

    .orange {
      color: $color-detail;
    }
  }
}

@media (min-width: 1200px) {
  .user-header {
    .user-header__banner {
      display: flex;
    }
  }
}
