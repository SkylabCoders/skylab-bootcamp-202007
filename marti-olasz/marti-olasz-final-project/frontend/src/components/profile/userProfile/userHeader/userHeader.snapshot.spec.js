import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { ProfileHeader } from './userHeader';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow } from 'enzyme';
Enzyme.configure({ adapter: new Adapter() });

describe('ProfileHeader snapshot', () => {
  const user = {
    _id: 1,
    photo: 'photo',
    banner: 'banner'
  };

  it('Should match without editInfo', () => {
    const editInfo = {};
    const image = {
      identifier: 'banner',
      path: 'path/'
    };

    const tree = renderer.create(
      <ProfileHeader user={user} editInfo={editInfo} image={image} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('Should match with editInfo', () => {
    const editInfo = { user: 'user', bio: 'bio' };
    const image = {
      identifier: 'profile-photo',
      path: 'path/'
    };

    const tree = renderer.create(
      <ProfileHeader user={user} editInfo={editInfo} image={image} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  const editInfo = { user: 'user', bio: 'bio' };
  const dispatch = jest.fn();

  it('Should call dispatch when click gear', () => {
    const document = shallow(
      <ProfileHeader user={user} editInfo={editInfo} dispatch={dispatch} />
    );
    const button = document.find('.edit__gear');
    button.simulate('click', {});

    expect(dispatch.call).truthy;
  });

  it("Should don't call dispatch when click gear", () => {
    const editInfo = {};
    const document = shallow(
      <ProfileHeader user={user} editInfo={editInfo} dispatch={dispatch} />
    );
    const button = document.find('.edit__gear');
    button.simulate('click', {});

    expect(dispatch.call).truthy;
  });
});
