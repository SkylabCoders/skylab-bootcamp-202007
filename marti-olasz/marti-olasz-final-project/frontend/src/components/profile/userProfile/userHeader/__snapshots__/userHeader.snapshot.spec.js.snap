// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileHeader snapshot Should match with editInfo 1`] = `
<section
  className="user-header"
>
  <img
    alt="user"
    className="user-header__photo"
    onClick={[Function]}
    src="photo"
  />
  <input
    name="file"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="user-header__banner"
  >
    <img
      alt="Banner"
      className="banner__img"
      onClick={[Function]}
      src="banner"
    />
    <input
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
  </div>
  <div
    className="user-header__edit"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root edit__gear orange"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
        transform="scale(1.2, 1.2)"
      />
    </svg>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root edit__save"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
      />
    </svg>
  </div>
</section>
`;

exports[`ProfileHeader snapshot Should match without editInfo 1`] = `
<section
  className="user-header"
>
  <img
    alt="user"
    className="user-header__photo"
    onClick={[Function]}
    src="photo"
  />
  <input
    name="file"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="user-header__banner"
  >
    <img
      alt="Banner"
      className="banner__img"
      onClick={[Function]}
      src="banner"
    />
    <input
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
  </div>
  <div
    className="user-header__edit"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root edit__gear white"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
        transform="scale(1.2, 1.2)"
      />
    </svg>
  </div>
</section>
`;
